lit_tests/kernel/wave/expansion.py

```diff
@@ -0,0 +1,342 @@
+    The graphs are printed first in the torch printing format and then using
+    our custom node format.
+    """
+    # The root graph is at the back so we print the subgraphs in reverse order
+    for subgraph in reversed(list(trace.region_graph.subgraphs.values())):
+        print(subgraph)
+
+
+# Input sizes
+M = tkl.sym.M
+N = tkl.sym.N
+K = tkl.sym.K
+
+# Workgroup tile sizes
+BLOCK_M = tkl.sym.BLOCK_M
+BLOCK_N = tkl.sym.BLOCK_N
+BLOCK_K = tkl.sym.BLOCK_L
+
+# Address space (for GPU, shared(1) or global(0))
+ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+
+
+@tkw.wave_trace_only()
+def read_write_same_size(
+    a: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f16],
+    c: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f32],
+):
+    a_reg = tkw.read(a, elements_per_thread=4)
+    tkw.write(a_reg, c, elements_per_thread=4)
+
+
+@run
+def test_read_write_equal_sizes():
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [tkw.HardwareConstraint(threads_per_wave=64, waves_per_block=[1, 1])]
+
+    with tk.gen.TestLaunchContext(
+        {
+            BLOCK_M: 32,
+            BLOCK_N: 32,
+        }
+    ):
+        graph = read_write_same_size()
+        IndexingContext.current().finalize()
+        expand_graph(graph, constraints)
+        print_trace(graph)
+        # CHECK: %a
+        # CHECK: %c
+        # CHECK: %read
+        # CHECK-SAME: (%a, 4)
+        # CHECK: %read_1_1
+        # CHECK-SAME: (%a, 4)
+        # CHECK: %read_1_0
+        # CHECK-SAME: (%a, 4)
+        # CHECK: %read_0_1
+        # CHECK-SAME: (%a, 4)
+        # CHECK: %write_0_0
+        # CHECK-SAME: (%read, %c, 4)
+        # CHECK: %write_1_1
+        # CHECK-SAME: (%read_1_1, %c, 4)
+        # CHECK: %write_1_0
+        # CHECK-SAME: (%read_1_0, %c, 4)
+        # CHECK: %write_0_1
+        # CHECK-SAME: (%read_0_1, %c, 4)
+
+        # CHECK: -----
+
+
+@tkw.wave_trace_only()
+def read_write_different_dims(
+    a: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f16],
+    c: tkl.Memory[M, K, ADDRESS_SPACE, tkl.f32],
+):
+    a_reg = tkw.read(a, elements_per_thread=4)
+    tkw.write(a_reg, c, elements_per_thread=4)
+
+
+@run
+def test_read_write():
+    constraints: list[tkw.Constraint] = [tkw.WorkgroupConstraint(M, BLOCK_M, 0)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 1)]
+    constraints += [tkw.WorkgroupConstraint(K, BLOCK_K, 2)]
+    constraints += [
+        tkw.HardwareConstraint(threads_per_wave=64, waves_per_block=[1, 1, 1])
+    ]
+    with tk.gen.TestLaunchContext(
+        {
+            BLOCK_M: 32,
+            BLOCK_N: 32,
+            BLOCK_K: 32,
+        }
+    ):
+        graph = read_write_different_dims()
+        IndexingContext.current().finalize()
+        expand_graph(graph, constraints)
+        print_trace(graph)
+        # CHECK: %a
+        # CHECK: %c
+        # CHECK: %read_0_0_0
+        # CHECK-SAME: (%a, 4)
+        # CHECK: %read_1_0_0
```

Comment:
if all the tile sizes are 32 and mma size is 16, shouldnt there be 4 reads and 4 writes?